### ISSUE TITLE
React-Native project can not compile npm pkcs11js module for iOS 

### ISSUE THREAD

**Issue by Vanclief** [janeasystems/nodejs-mobile#134]
* **Project**: nodejs-mobile-react-native
* **Version**: 0.3.0
* **Mobile device**: iOS iPhone 6 Simulator
* **Mobile OS and version**: N/A
* **Development Node.js**: 10.11.0
* **Development platform**: Darwin Kernel 17.7.0

**Expected behavior:**
I can install a native [pkcs11js](https://github.com/PeculiarVentures/pkcs11js) npm module and then compile my react native project for iOS.

**Actual behavior:** 
Compiler fails when trying to build the application after installing [pkcs11js](https://github.com/PeculiarVentures/pkcs11js)

**How to reproduce:**

1. Create new react native project with support for native libraries
2. Add the nodejs-mobile-react-native library
3. Install pkcs11js in the nodejs-projects folder: `npm install pkcs11js`
4. Compile the iOS project.
5. Error:

```
ake: *** [Release/obj.target/pkcs11/src/main.o] Error 1                                                                                                                                                   [74/5643]

gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-gyp/lib/build.js:258:23)
gyp ERR! stack     at ChildProcess.emit (events.js:182:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:240:12)

gyp ERR! System Darwin 17.7.0
gyp ERR! command "/usr/local/Cellar/node/10.11.0/bin/node" "/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js
gyp ERR! node -v v10.11.0
gyp ERR! node-gyp -v v0.2.0

gyp ERR! not ok

npm verb lifecycle pkcs11js@1.0.16~install: unsafe-perm in lifecycle true
npm verb lifecycle pkcs11js@1.0.16~install: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaC
acao.app/nodejs-project/node_modules/pkcs11js/node_modules/.bin:/Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/.bin:/Applications/Xc
ode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin:/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/local/bin:/Applications/Xcode.app/Contents/Developer/Toolc
hains/XcodeDefault.xctoolchain/usr/libexec:/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/usr/bin:/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.plat
form/Developer/usr/local/bin:/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/usr/bin:/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/usr/local/bin:/Appl
ications/Xcode.app/Contents/Developer/usr/bin:/Applications/Xcode.app/Contents/Developer/usr/local/bin:/Applications/Xcode.app/Contents/Developer/Tools:/Users/franco/.fastlane/bin:/bin:/Users/franco/.rvm/bin:/Use
rs/franco/.bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/usr/local/MacGPG2/bin:/Users/franco/anaconda3/bin:/home/vanclief/go/bin:/Users/franco/.fzf/bin
npm verb lifecycle pkcs11js@1.0.16~install: CWD: /Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js
npm info lifecycle pkcs11js@1.0.16~install: Failed to exec install script

npm verb stack Error: pkcs11js@1.0.16 install: `node-gyp rebuild`
npm verb stack Exit status 1

npm verb stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:301:16)

npm verb stack     at EventEmitter.emit (events.js:182:13)

npm verb stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
npm verb stack     at ChildProcess.emit (events.js:182:13)
npm verb stack     at maybeClose (internal/child_process.js:962:16)

npm verb stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
npm verb pkgid pkcs11js@1.0.16
npm verb cwd /Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project

npm verb Darwin 17.7.0

npm verb argv "/usr/local/Cellar/node/10.11.0/bin/node" "/usr/local/bin/npm" "--verbose" "rebuild" "--build-from-source"
npm verb node v10.11.0

npm verb npm  v6.4.1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pkcs11js@1.0.16 install: `node-gyp rebuild`

npm ERR! Exit status 1

npm ERR!

npm ERR! Failed at the pkcs11js@1.0.16 install script.

npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm verb exit [ 1, true ]

npm timing npm Completed in 127383ms



npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/franco/.npm/_logs/2018-10-11T20_44_37_018Z-debug.log



** BUILD FAILED **
```

**Comment by jaimecbernardo** [comment]
Hi @Vanclief ,
The specific compile errors that kept the module from building should be in the output as well, above the part that you pasted.
I hope this is helpful.

**Comment by Vanclief** [comment]
Hey @jaimecbernardo, thanks for the heads up. This is the relevant error log:

```

> pkcs11js@1.0.16 install /Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js
> node-gyp rebuild


gyp info it worked if it ends with ok

gyp verb cli [ '/usr/local/Cellar/node/10.11.0/bin/node',

gyp verb cli   '/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-gyp/bin/node-gyp.js',

gyp verb cli   'rebuild' ]

gyp info using node-gyp@0.2.0

gyp info using node@10.11.0 | darwin | x64

gyp verb command rebuild []

gyp verb command clean []

gyp verb clean removing "build" directory

gyp verb command configure []

gyp verb check python checking for Python executable "python2" in the PATH

gyp verb `which` succeeded python2 /usr/local/bin/python2

gyp verb check python version `/usr/local/bin/python2 -c "import platform; print(platform.python_version());"` returned: "2.7.15\n"

gyp verb get node dir compiling against specified --nodedir dev files: /Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-react-native/ios/libnode

gyp verb build dir attempting to create "build" dir: /Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js/build

gyp verb build dir "build" dir needed to be created? /Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js/build

gyp verb build/config.gypi creating config file

gyp verb build/config.gypi writing out config file: /Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js/build/config.gypi

gyp verb config.gypi checking for gypi file: /Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js/config.gypi

gyp verb common.gypi checking for gypi file: /Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js/common.gypi

gyp verb gyp gyp format was not specified; forcing "make"

gyp info spawn /usr/local/bin/python2

gyp info spawn args [ '/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-gyp/gyp/gyp_main.py',

gyp info spawn args   'binding.gyp',

gyp info spawn args   '-f',

gyp info spawn args   'make-ios',

gyp info spawn args   '-I',

gyp info spawn args   '/Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js/build/config.gypi',

gyp info spawn args   '-I',

gyp info spawn args   '/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-gyp/addon.gypi',

gyp info spawn args   '-I',

gyp info spawn args   '/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-react-native/ios/libnode/include/node/common.gypi',

gyp info spawn args   '-Dlibrary=shared_library',

gyp info spawn args   '-Dvisibility=default',

gyp info spawn args   '-Dnode_root_dir=/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-react-native/ios/libnode',

gyp info spawn args   '-Dnode_gyp_dir=/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-gyp',

gyp info spawn args   '-Dnode_lib_file=/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-react-native/ios/libnode/$(Configuration)/node.lib',

gyp info spawn args   '-Dmodule_root_dir=/Users/franco/Cacao_repos/react-app/ios/build/Build/Products/Debug-iphonesimulator/MonedaCacao.app/nodejs-project/node_modules/pkcs11js',

gyp info spawn args   '-Dnode_engine=chakracore',

gyp info spawn args   '--depth=.',

gyp info spawn args   '--no-parallel',

gyp info spawn args   '--generator-output',

gyp info spawn args   'build',

gyp info spawn args   '-Goutput_dir=.' ]

gyp verb command build []

gyp verb build type Release

gyp verb architecture x64

gyp verb node dev dir /Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-react-native/ios/libnode

gyp verb `which` succeeded for `make` /Applications/Xcode.app/Contents/Developer/usr/bin/make

gyp info spawn make

gyp info spawn args [ 'V=1', 'BUILDTYPE=Release', '-C', 'build' ]

  c++ '-DNODE_GYP_MODULE_NAME=pkcs11' '-DUSING_UV_SHARED=1' '-DUSING_V8_SHARED=1' '-DV8_DEPRECATION_WARNINGS=1' '-DNODE_ENGINE="chakracore"' '-DNODE_ENGINE_CHAKRACORE' '-D_DARWIN_USE_64_BIT_INODE=1' '-D_LARGEFILE
_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-react-native/ios/libnode/include/node -I/Users/franco/Cacao_repos/react-app/node_modu
les/nodejs-mobile-react-native/ios/libnode/src -I/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-react-native/ios/libnode/deps/uv/include -I/Users/franco/Cacao_repos/react-app/node_modules/nodejs-m
obile-react-native/ios/libnode/deps/chakrashim/include -I../../nan -I../includes  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator11.4.sdk -Os
 -gdwarf-2 -mios-simulator-version-min=9.0 -arch x86_64 -Wall -Wendif-labels -W -Wno-unused-parameter -std=gnu++0x -stdlib=libc++ -fno-rtti -fno-exceptions -fno-threadsafe-statics -fno-strict-aliasing -MMD -MF ./
Release/.deps/Release/obj.target/pkcs11/src/main.o.d.raw   -c -o Release/obj.target/pkcs11/src/main.o ../src/main.cpp

In file included from ../src/main.cpp:1:

../../nan/nan.h:139:9: warning: 'TYPE_CHECK' macro redefined [-Wmacro-redefined]

#define TYPE_CHECK(T, S)                                                       \

        ^

/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-react-native/ios/libnode/include/node/v8.h:85:9: note: previous definition is here

#define TYPE_CHECK(T, S)                                       \

        ^

In file included from ../src/main.cpp:4:

In file included from ../src/const.h:17:

In file included from ../src/pkcs11/param.h:5:

../src/pkcs11/v8_convert.h:23:42: error: cannot use 'throw' with exceptions disabled

        virtual void FromV8(Local<Value> obj) { THROW_ERROR("Not implemented", NULL); }

                                                ^

../src/pkcs11/error.h:36:6: note: expanded from macro 'THROW_ERROR'

     throw Scoped<Error>(new Error(msg, stack, __FUNCTION__, __LINE__))

     ^

In file included from ../src/main.cpp:4:

In file included from ../src/const.h:17:

In file included from ../src/pkcs11/param.h:5:

../src/pkcs11/v8_convert.h:24:33: error: cannot use 'throw' with exceptions disabled

        virtual Local<Object> ToV8() { THROW_ERROR("Not implemented", NULL); }

                                       ^

../src/pkcs11/error.h:36:6: note: expanded from macro 'THROW_ERROR'

     throw Scoped<Error>(new Error(msg, stack, __FUNCTION__, __LINE__))

     ^

1 warning and 2 errors generated.

make: *** [Release/obj.target/pkcs11/src/main.o] Error 1

gyp ERR! build error
```

My uneducated guess is this can be solved by allowing exceptions or by stripping the pkcs11 library from all error throws, but I am sure you will have a better idea how this can be worked around.

Again, thanks for all the help. This library is awesome and is allowing us to build our product in a way that otherwise it would not be possible. I hope I can contribute back somehow.

**Comment by jaimecbernardo** [comment]
Hi @Vanclief ,
Indeed the issue seems to be about the exceptions.

By default, C++ exceptions are disabled for modules, which the `pkcs11js` plugin overrides for macOS : https://github.com/PeculiarVentures/pkcs11js/blob/f6f46336994f3b7cc04a9714717bc9ba32ee0c4f/binding.gyp#L38

This overrides this setting from nodejs's `common.gypi`: https://github.com/janeasystems/nodejs-mobile-react-native/blob/ed727edea17e8a9e1a85cef3413becc83b8a0328/ios/libnode/include/node/common.gypi#L517

In `pkcs11js`'s `binding.gyp`, a section for iOS could be created to override the settings from `common.gypi`, like the it's being done for macOS: https://github.com/janeasystems/nodejs-mobile-react-native/blob/ed727edea17e8a9e1a85cef3413becc83b8a0328/ios/libnode/include/node/common.gypi#L443

Thank you for feedback.
One of the ways for contributing would be to open a "build workaround" issue in https://github.com/janeasystems/nodejs-mobile-module-compat once you get the plugin to work, explaining what was needed to change, and opening a PR to https://github.com/PeculiarVentures/pkcs11js with the new iOS section in `binding.gyp`. This way the `pkcs11js` plugin would eventually be compatible with ios out of the box.
Thank you, in advance.

**Comment by Vanclief** [comment]
Hey @jaimecbernardo,

Excellent, thanks for pointing me in the right direction, I will proceed to get the build working, create the workaround and make the PR.

I also noticed that `pkcs11js` is not building for Android either, seems it requires a `-fPIC` flag. Should this also be edited on the `pkcs11js` `biding.gyp` file, or in https://github.com/janeasystems/nodejs-mobile-react-native/blob/ed727edea17e8a9e1a85cef3413becc83b8a0328/android/libnode/include/node/common.gypi?

```
2 warnings generated.
  /Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/arm-linux-androideabi-clang++ -shared -rdynamic -fPIC  -Wl,-soname=pkcs11.node -o Release/obj.target/pkcs11.node
 -Wl,--start-group Release/obj.target/pkcs11/src/main.o Release/obj.target/pkcs11/src/dl.o Release/obj.target/pkcs11/src/const.o Release/obj.target/pkcs11/src/pkcs11/error.o Release/obj.target/pkcs11/src/pkcs11/v
8_convert.o Release/obj.target/pkcs11/src/pkcs11/template.o Release/obj.target/pkcs11/src/pkcs11/mech.o Release/obj.target/pkcs11/src/pkcs11/param.o Release/obj.target/pkcs11/src/pkcs11/param_aes.o Release/obj.ta
rget/pkcs11/src/pkcs11/param_rsa.o Release/obj.target/pkcs11/src/pkcs11/param_ecdh.o Release/obj.target/pkcs11/src/pkcs11/pkcs11.o Release/obj.target/pkcs11/src/async.o Release/obj.target/pkcs11/src/node.o -Wl,--
end-group -llog /Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-react-native/android/libnode//bin/armeabi-v7a/libnode.so
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/error.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/error.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/template.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/template.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/template.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/mech.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/param.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/param.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/param_aes.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/param_rsa.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/param_ecdh.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/pkcs11.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/pkcs11/pkcs11.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/async.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/async.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/async.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/async.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/async.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/async.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.targe[36/6165]
src/async.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/async.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/node.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/node.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/node.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/node.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/node.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/node.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/node.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
/Users/franco/Cacao_repos/react-app/android/build/standalone-toolchains/arm-linux-androideabi/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: Release/obj.target/pkcs11/
src/node.o: requires unsupported dynamic reloc R_ARM_REL32; recompile with -fPIC
clang70++: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Release/obj.target/pkcs11.node] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-gyp/lib/build.js:258:23)
gyp ERR! stack     at ChildProcess.emit (events.js:182:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:240:12)
gyp ERR! System Darwin 17.7.0
gyp ERR! command "/usr/local/Cellar/node/10.11.0/bin/node" "/Users/franco/Cacao_repos/react-app/node_modules/nodejs-mobile-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/franco/Cacao_repos/react-app/android/build/nodejs-native-assets-temp-build/nodejs-native-assets-armeabi-v7a/nodejs-project/node_modules/pkcs11js
gyp ERR! node -v v10.11.0
gyp ERR! node-gyp -v v0.2.0
gyp ERR! not ok
npm verb lifecycle pkcs11js@1.0.16~install: unsafe-perm in lifecycle true
npm verb lifecycle pkcs11js@1.0.16~install: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/Users/franco/Cacao_repos/react-app/android/build/nodejs-native-assets-temp-build/nodejs-n
ative-assets-armeabi-v7a/nodejs-project/node_modules/pkcs11js/node_modules/.bin:/Users/franco/Cacao_repos/react-app/android/build/nodejs-native-assets-temp-build/nodejs-native-assets-armeabi-v7a/nodejs-project/no
de_modules/.bin:/Users/franco/Cacao_repos/react-app/android/../nodejs-assets/nodejs-project/node_modules/.bin:/Users/franco/.fastlane/bin:/bin:/Users/franco/.rvm/bin:.git/safe/../../bin:/Users/franco/.bin:/usr/lo
cal/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/usr/local/MacGPG2/bin:/Users/franco/anaconda3/bin:/home/vanclief/go/bin:/Users/franco/Library/Android/sdk/tools:/Users/franco/Library/Android/sdk/platform-tools:/
Users/franco/.fzf/bin
npm verb lifecycle pkcs11js@1.0.16~install: CWD: /Users/franco/Cacao_repos/react-app/android/build/nodejs-native-assets-temp-build/nodejs-native-assets-armeabi-v7a/nodejs-project/node_modules/pkcs11js
npm info lifecycle pkcs11js@1.0.16~install: Failed to exec install script
npm verb stack Error: pkcs11js@1.0.16 install: `node-gyp rebuild`
npm verb stack Exit status 1
npm verb stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:301:16)
npm verb stack     at EventEmitter.emit (events.js:182:13)
npm verb stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
npm verb stack     at ChildProcess.emit (events.js:182:13)
npm verb stack     at maybeClose (internal/child_process.js:962:16)
npm verb stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
npm verb pkgid pkcs11js@1.0.16
npm verb cwd /Users/franco/Cacao_repos/react-app/android/build/nodejs-native-assets-temp-build/nodejs-native-assets-armeabi-v7a/nodejs-project
npm verb Darwin 17.7.0
npm verb argv "/usr/local/Cellar/node/10.11.0/bin/node" "/usr/local/bin/npm" "--verbose" "rebuild" "--build-from-source"
npm verb node v10.11.0
npm verb npm  v6.4.1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pkcs11js@1.0.16 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the pkcs11js@1.0.16 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm verb exit [ 1, true ]
npm timing npm Completed in 23775ms

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/franco/.npm/_logs/2018-10-17T04_58_00_960Z-debug.log
:nodejs-mobile-react-native:BuildNpmModulesarmeabi-v7a FAILED

FAILURE: Build failed with an exception.
```

**Comment by Vanclief** [comment]
Interestingly enough I am seeing the `common.gypi` for android already has the `-fPIC` flag: https://github.com/janeasystems/nodejs-mobile-react-native/blob/ed727edea17e8a9e1a85cef3413becc83b8a0328/android/libnode/include/node/common.gypi#L404. But still, it is not working as expected.  This is weird.

**Comment by jaimecbernardo** [comment]
Hi @Vanclief ,
It seems the flag is being added to the link command, if you check the log you posted. It's possible that it's not present in the previous compile commands. You may have more information in the previous log messages.

**Comment by Vanclief** [comment]
Hey @jaimecbernardo, thats exactly what I noticed, which makes it harder to pinpoint the cause. I have tried with different flags with no success. It is also interesting this is happening for multiple modules, which normally compile by themselves. This is the full error log: https://gist.github.com/Vanclief/518de77b69aea8cca29e64bdb24c829e

**Comment by Vanclief** [comment]
Last comment was a little bit of topic, and I managed to get it working. So I am closing this thread.

If you find this thread, the solution is here:
https://github.com/janeasystems/nodejs-mobile-module-compat/issues/4

**Comment by jaimecbernardo** [comment]
For this issue, a change to add `-fPIC` was added to `nodejs-mobile-gyp`: https://github.com/janeasystems/nodejs-mobile-gyp/commit/fa19d39733f08cf121479de8a0a6985ebc815e8a

This change was included with `nodejs-mobile-gyp` version `0.2.1`
